crypto/s2n_rsa_pss.c

```diff
@@ -176,14 +177,22 @@ static int s2n_rsa_pss_keys_match(const struct s2n_pkey *pub, const struct s2n_p
 
 static int s2n_rsa_pss_key_free(struct s2n_pkey *pkey)
 {
-    /* This object does not own the reference to the key --
-     * s2n_pkey handles it. */
+    POSIX_ENSURE_REF(pkey);
+    struct s2n_rsa_key *rsa_key = &pkey->key.rsa_key;
+    if (rsa_key->rsa == NULL) {
+        return S2N_SUCCESS;
+    }
 
-    return 0;
+    /* Safety: freeing the key owned by this object */
+    RSA_free(s2n_unsafe_rsa_get_non_const(rsa_key));
+    rsa_key->rsa = NULL;
+
+    return S2N_SUCCESS;
 }
 
 int s2n_evp_pkey_to_rsa_pss_public_key(struct s2n_rsa_key *rsa_key, EVP_PKEY *pkey) {
-    RSA *pub_rsa_key = EVP_PKEY_get0_RSA(pkey);
+    const RSA *pub_rsa_key = EVP_PKEY_get1_RSA(pkey);
+    POSIX_ENSURE_REF(pub_rsa_key);
```

Comment:
I just noticed that you're swapping EVP_PKEY_get0_RSA for EVP_PKEY_get1_RSA for RSA-PSS. That does introduce a new (probably small?) allocation. Are you just doing this for consistency? What's the reasoning?

Reply:
Consistency is the main reason. I have listed a few differences below and since ossl functions are quite complicated I think its better to use a smaller surface area and behavior (only get1) rather than optimize for a single allocation.

Some differences between get1 and get0:
- get1 is used in rsa and ecdsa and code is more consistent
- ossl3 changed get0 to return const while get1 remains non-const
- get1 requires a call to free while get0 doesnt
- get0 is not available in ossl1.0.2


Future work:
If we are able to deprecate ossl1.0.2, we should replace usage with PKEY and then use get0 where we need the underlying key.

Reply:
Regarding allocation: btw `get1` doe NOT actually do an extra allocation compared with `get0`. `get1` simply does a `ref_count++` and therefore requires a call to free.

https://github.com/openssl/openssl/blob/31b7f23d2f958491d46c8a8e61c2b77b1b546f3e/crypto/evp/p_legacy.c#L53

Reply:
I asked because most of those difference don't apply to RSA-PSS, or are in get0's favor. We want a const, the extra free call is more work, and RSA-PSS isn't available in Openssl-1.0.2 either :)

But consistency seems like an okay reason to switch, especially if there's no allocation cost!